image/manifest_test.go

```diff
@@ -48,3 +64,60 @@ func TestUnpackLayerDuplicateEntries(t *testing.T) {
 		t.Fatalf("Expected to fail with duplicate entry, got %v", err)
 	}
 }
+func TestUnpackLayerRemovePartialyUnpackedFile(t *testing.T) {
+	tmp1, err := ioutil.TempDir("", "test-layer")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(tmp1)
+	err = os.MkdirAll(filepath.Join(tmp1, "blobs", "sha256"), 0700)
+	if err != nil {
+		t.Fatal(err)
+	}
+	tarfile := filepath.Join(tmp1, "blobs", "sha256", "test.tar")
```

Comment:
On Wed, Sep 07, 2016 at 11:45:29PM -0700, Lei Jitang wrote:
> I don't know if I got your point, I did it similarly like you said
> at the beginning, put the `tar` file under `tmp1` and then copy it
> to `blobs/sha256/` after renaming it to `hex hash`.  BUt to avoid a
> unnecessary copy, so I put the `tar` file to the destination dir at
> the beginning and rename it after calculating hash.

You can use rename no matter what the source and destination are, and
only have to pay for a copy if they are on different filesystems.  In
this case, tmp1/…, {tmp1}/tmp/… and {tmp1}/blobs/sha256/… are all in
the same filesystem, so rename will be cheap in all cases.